src/cascadia/TerminalSettingsModel/Profile.cpp

```diff
@@ -533,5 +538,154 @@ Json::Value Profile::ToJson() const
+    //             that each layer in the recursion only has keys seen leading up to it
+    // Return Value:
+    // - The resolved value for key
+    std::wstring ResolveEnvironmentVariableValue(const std::wstring& key, std::map<std::wstring, VariablePair>& envMap, std::list<std::wstring> seenKeys = {})
+    {
+        // First check if the key is in our environment variable map
+        auto it = envMap.find(key);
+        if (it != envMap.end())
+        {
+            if (!it->second.resolvedValue.empty())
+            {
+                // key has already been resolved, so just return it
+                return it->second.resolvedValue;
+            }
+            if (std::find(seenKeys.cbegin(), seenKeys.cend(), key) != seenKeys.end())
+            {
+                // If we've attempted to resolve this key already then we're circularly resolving keys and this is
+                // an error, not aborting here would result in a dead-lock. e.g.:
+                // {
+                //    'KEY_1': '${env:KEY_2}',
+                //    'KEY_2': '${env:KEY_1}'
+                // }
+                std::wstringstream error;
+                error << L"Self referencing keys in environment settings: ";
+                for (auto seenKey : seenKeys)
+                {
+                    error << L"'" << seenKey << L"', ";
+                }
+                error << L"'" << key << L"'";
+                throw winrt::hresult_error(WEB_E_INVALID_JSON_STRING, winrt::hstring(error.str()));
+            }
+            static const std::wregex parameterRegex(LR"(\$\{env:(.*?)\})");
+            auto& value = it->second;
+            auto textIter = value.rawValue.cbegin();
+            std::wsmatch regexMatch;
+            while (std::regex_search(textIter, value.rawValue.cend(), regexMatch, parameterRegex))
+            {
+                if (regexMatch.size() != 2)
+                {
+                    std::wstringstream error;
+                    error << L"Unexpected regex match count '" << regexMatch.size()
+                          << L"' (expected '2') when matching '" << std::wstring(textIter, value.rawValue.cend()) << L"'";
+                    throw winrt::hresult_error(WEB_E_INVALID_JSON_STRING, winrt::hstring(error.str()));
+                }
+                value.resolvedValue += std::wstring(textIter, regexMatch[0].first);
+                const auto referencedKey = regexMatch[1];
+                if (referencedKey == key)
+                {
+                    // This key references itself, this is another dead-lock condition. e.g.:
+                    // {
+                    //   'KEY_1': 'abc${env:KEY_1}def'
+                    // }
+                    std::wstringstream error;
+                    error << L"Self referencing key '" << key << L"' in environment settings";
+                    throw winrt::hresult_error(WEB_E_INVALID_JSON_STRING, winrt::hstring(error.str()));

```

Comment:
I expect users will complain if they cannot use this for appending to `PATH`.

Reply:
That you for this (and your other comments), I dropped the ball here..